Feat/nextflow run for bioimaging (#680)
nextflow-io/training: docs/nf4_science/imaging/01_basics.md

> # Part 1: Run basic operations

Author: There are a few minor customizations here. I dropped the section 0 (running things using bash alone). I'm also customizing the "Next Steps" section. Going to leave this as a content copy for now, but we'll figure out a better way for custom paths in the future.

For each one, it will stage the necessary inputs, execute the relevant instruction(s) and write outputs and log files within that one directory, which is named automatically using a hash in order to make it unique.

All of these task directories will live under a directory called `work` within your current directory (where you're running the command).

That may sound confusing, so let's see what that looks like in practice.

Going back to the console output for the workflow we ran earlier, we had this line:

```console title="Excerpt of command output" linenums="6"
[a3/7be2fa] sayHello | 1 of 1 ✔
```

See how the line starts with `[a3/7be2fa]`?
That is a truncated form of the task directory path for that one process call, and tells you where to find the output of the `sayHello` process call within the `work/` directory path.

You can find the full path by typing the following command (replacing `a3/7be2fa` with what you see in your own terminal) and pressing the tab key to autocomplete the path or adding an asterisk:

```bash

> ls work/a3/7be2fa*

Reviewer: ```suggestion
tree -a work/a3/7be2fa*
```
you switch a lot between 'ls' and 'tree' - this tutorial might contain a lot of non-tech savvy, folks, i would say just pick one or the other or you may end up losing folks out of confusion

```

This should yield the full path directory path: `work/a3/7be2fa7be2fad5e71e5f49998f795677fd68`

Let's take a look at what's in there.

!!! Tip

    If you browse the contents of the task subdirectory in the VSCode file explorer, you'll see all the files right away.
    However, the log files are set to be invisible in the terminal, so if you want to use `ls` or `tree` to view them, you'll need to set the relevant option for displaying invisible files.

    ```bash
    tree -a work
    ```

The exact subdirectory names will be different on your system.

<details>
  <summary>Directory contents</summary>

```console title="work/"
work
└── a3
    └── 7be2fad5e71e5f49998f795677fd68
        ├── .command.begin
        ├── .command.err
        ├── .command.log
        ├── .command.out
        ├── .command.run
        ├── .command.sh
        ├── .exitcode
        └── output.txt
```

</details>

You should immediately recognize the `output.txt` file, which is in fact the original output of the `sayHello` process that got published to the `results` directory.
If you open it, you will find the `Hello World!` greeting again.

<details>

>   <summary>File contents</summary>

Reviewer: ```suggestion
<summary>File contents of output.txt</summary>
```

- **`.command.out`**: Regular output (`stdout`) by the process call
- **`.command.run`**: Full script run by Nextflow to execute the process call
- **`.command.sh`**: The command that was actually run by the process call
- **`.exitcode`**: The exit code resulting from the command

The `.command.sh` file is especially useful because it shows you the main command Nextflow executed not including all the bookkeeping and task/environment setup.

<details>
  <summary>File contents</summary>

```console title="work/a3/7be2fa7be2fad5e71e5f49998f795677fd68/command.sh" linenums="1"
#!/bin/bash -ue
echo 'Hello World!' > output.txt

```

</details>

!!! Tip

    When something goes wrong and you need to troubleshoot what happened, it can be useful to look at the `command.sh` script to check exactly what command Nextflow composed based on the workflow instructions, variable interpolation and so on.

### 1.4. Optional exercise: re-run with different greetings

Try re-running the workflow a few times with different values for the `--greeting` argument, then look at both the contents of the `results/` directory and the task directories.

> Observe how the outputs and logs of isolated task directories are preserved, whereas the contents of the `results` directory are overwritten by the output of subsequent executions.

Reviewer: ```suggestion
Observe how the outputs and logs of isolated task directories are preserved, whereas the contents of the `results` directory are unique and contain the output of subsequent executions.
```

<details>
  <summary>Command output</summary>

```console linenums="1"
 N E X T F L O W   ~  version 25.04.3

Launching `hello-world.nf` [tiny_noyce] DSL2 - revision: c33d41f479

[a3/7be2fa] process > sayHello [100%] 1 of 1, cached: 1 ✔
```

</details>

Look for the `cached:` bit that has been added in the process status line (line 5), which means that Nextflow has recognized that it has already done this work and simply reused the result from the previous successful run.

You can also see that the work subdirectory hash is the same as in the previous run.
Nextflow is literally pointing you to the previous execution and saying "I already did that over there."

!!! Tip

    When your re-run a pipeline with `resume`, Nextflow does not overwrite any files written to a `publishDir` directory by any process call that was previously run successfully.

### 3.2. Inspect the log of past executions

Whenever you launch a nextflow workflow, a line gets written to a log file called `history`, under a hidden directory called `.nextflow` in the current working directory.

A more convenient way to access this information is to use the `nextflow log` command.

```bash
nextflow log
```

> 

Reviewer 2: [![suggestion](https://img.shields.io/badge/suggestion-3B82F6)](https://pullpo.io/cc?l=suggestion) 
Maybe include a screenshot / codeblock from `nextflow log` here for people to understand what it looks like when only reading through the training?

This will output the contents of the log file to the terminal, showing you the timestamp, run name, status, and full command line for every Nextflow run that has been launched from within the current working directory.

### 3.3. Delete older work directories

During the development process, you'll typically run your draft pipelines a large number of times, which can lead to an accumulation of very many files across many subdirectories.
Since the subdirectories are named randomly, it is difficult to tell from their names what are older vs. more recent runs.

Nextflow includes a convenient `clean` subcommand that can automatically delete the work subdirectories for past runs that you no longer care about, with several [options](https://www.nextflow.io/docs/latest/reference/cli.html#clean) to control what will be deleted.

> You can use the Nextflow log to look up a run based on its timestamp and/or command line, then use `nextflow clean -before <run_name> -f` to delete work directories from earlier runs.

Reviewer: ```suggestion
You can use the Nextflow log to look up a run based on its timestamp and/or command line, then use `nextflow clean -before tiny_noyce -f` (the run name from the previous example) to delete work directories from earlier runs.
```

Author: I actually like the generic term better in this case.
